test/FlowChartSpec.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+module FlowChartSpec
+  ( spec,
+  )
+where
+
+import Algebra.Graph.Labelled as LG (edgeList, vertexList)
+import FlowChart
+import NeatInterpolation
+import Test.Hspec
+import Text.Megaparsec
+
+shouldHaveEdges :: Maybe Diagram -> [(Maybe Edge, Node, Node)] -> Expectation
+shouldHaveEdges = \x -> (\y -> shouldBe (edgeList . graph <$> x) (Just y))
+
+infix 1 `shouldHaveEdges`
+
+shouldHaveNodes :: Maybe Diagram -> [Node] -> Expectation
+shouldHaveNodes = \x -> (\y -> shouldBe (vertexList . graph <$> x) (Just y))
+
+infix 1 `shouldHaveNodes`
+
+spec :: Spec
+spec = do
+  describe "FlowChart" $ do
+    it "nodes of single node" $ do
+      let diagram =
+            [text|
+              flowchart TD
+                id
+            |]
+      parseMaybe pDiagram diagram
+        `shouldHaveNodes` [Node {nodeId = NodeId {unNodeId = "id"}, nodeShape = Nothing, nodeLabel = Nothing}]
+    it "edges of single node" $ do
+      let diagram =
+            [text|
+              flowchart TD
+                id
+            |]
+      parseMaybe pDiagram diagram
+        `shouldHaveEdges` []
+    it "nodes of single edge with two nodes" $ do
+      let diagram =
+            [text|
+              flowchart TD
+                start --> stop
+            |]
+      parseMaybe pDiagram diagram
+        `shouldHaveNodes` [ Node {nodeId = NodeId {unNodeId = "start"}, nodeShape = Nothing, nodeLabel = Nothing},
+                            Node {nodeId = NodeId {unNodeId = "stop"}, nodeShape = Nothing, nodeLabel = Nothing}
```

Comment:
Have you considered simplifying all of this using `lens` and [`IsString`](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-String.html#v:fromString)? You can effectively do something like this to replace this line:

```
  fromString "stop"
```

And with lens, if you are specifying custom props:

```
  fromString "stop" & nodeShape .~ Just someShape
```

See the Model type and `makeLenses` to see how lens is being used in Emanote: https://github.com/srid/emanote/blob/master/src/Emanote/Model/Type.hs